File: ios/AppPostItKeyboard/UsageTracker.swift
import Foundation

private let insertCountKey = "insert_count"
private let isPremiumKey = "is_premium"

/// Tracks how many times a saved post has actually been inserted into
/// another app via the keyboard -- this, not how many posts are saved, is
/// what counts against the free tier. Mirrors UsageTracker.kt on Android;
/// reads/writes the same shared JSON file (SharedState.swift) the Flutter
/// app's SharedStorage bridges into (AppGroupPlugin.swift).
final class UsageTracker {
    func getInsertCount() -> Int {
        SharedState.getInt(insertCountKey) ?? 0
    }

    func recordInsert() {
        SharedState.setInt(insertCountKey, getInsertCount() + 1)
    }
}

/// Mirrors PurchaseStatusReader.kt on Android -- reads the "is_premium"
/// flag the Flutter app writes after a successful purchase.
final class PurchaseStatusReader {
    func isPremium() -> Bool {
        SharedState.getBool(isPremiumKey) ?? false
    }
}

File: android/app/src/main/kotlin/com/apppostit/apppostit/ime/UsageTracker.kt
package com.apppostit.apppostit.ime

import android.content.Context

private const val INSERT_COUNT_KEY = "flutter.insert_count"

/**
 * Tracks how many times a saved post has actually been inserted into
 * another app via the keyboard -- this is what counts against the free
 * tier, not how many posts are saved. Stored in the same
 * SharedPreferences file the Flutter app reads, under the "flutter."
 * prefix its shared_preferences plugin expects when it later reads this
 * key back via SharedPreferencesAsync.
 */
class UsageTracker(context: Context) {
    private val prefs =
        context.getSharedPreferences("FlutterSharedPreferences", Context.MODE_PRIVATE)

    fun getInsertCount(): Int = prefs.getInt(INSERT_COUNT_KEY, 0)

    fun recordInsert() {
        prefs.edit().putInt(INSERT_COUNT_KEY, getInsertCount() + 1).apply()
    }
}
